src/librustc_driver/args/mod.rs

```diff
@@ -0,0 +1,184 @@
+use std::env;
+use std::ffi::OsString;
+use std::fs;
+use std::mem;
+use std::path::PathBuf;
+
+#[cfg(test)]
+mod tests;
+
+// Max number of recursive @file includes
+const MAX_RECURSION: usize = 8;
+
+/// This uses the same conventions as GNU ld's handling of args files:
+/// @file - Read command-line options from file. The options read are inserted in place of the
+/// original @file option. If file does not exist, or cannot be read, then the option will be
+/// treated literally, and not removed.
+///
+/// Options in file are separated by whitespace. A whitespace character may be included in an
+/// option by surrounding the entire option in either single or double quotes. Any character
+/// (including a backslash) may be included by prefixing the character to be included with a
+/// backslash. The file may itself contain additional @file options; any such options will be
+/// processed recursively.
+///
+/// See https://sourceware.org/binutils/docs/ld/Options.html#Options for the documentation of
+/// gnu-ld's @file parameters.
+pub struct ArgsStack {
+    base: env::ArgsOs,
+    stack: Vec<Box<dyn Iterator<Item = FileArg>>>,
+}
+
+impl ArgsStack {
+    pub fn new() -> Self {
+        ArgsStack { base: env::args_os(), stack: vec![] }
+    }
+
+    fn stack_next(&mut self) -> Option<FileArg> {
+        while let Some(top) = self.stack.last_mut() {
+            match top.next() {
+                Some(filearg) => return Some(filearg),
+                None => {
+                    let _ = self.stack.pop();
+                }
+            }
+        }
+
+        None
+    }
+}
+
+impl Iterator for ArgsStack {
+    type Item = OsString;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            let arg = self.stack_next().or_else(|| self.base.next().map(FileArg::from));
+            match arg {
+                None => return None,
+                Some(FileArg::Arg(arg)) => return Some(arg),
+                Some(FileArg::Path(path)) => {
+                    let file =
+                        if self.stack.len() >= MAX_RECURSION { None } else { fs::read(&path).ok() };
```

Comment:
It feels a bit error-prone to simply ignore `@files` that are "too deep". Maybe we can at least `eprintln` or something here? There's probably no good way to give a good error message here...

One possible "solution" is to just panic and bump MAX_RECURSION to some very large number like 1000 which would be relatively unreasonable.

Reply:
The documented error handling is pretty awkward - it also just includes the literal `@file` string as an arg if it can't open/read the file. The recursion handling is at least consistent with that low standard.

I'm happy to scatter some `eprintln`s (if that's the best that's available here) around to make noise about things going wrong.